### PR TITLE
ARROW-11502: [C++] Optimize Arrow ByteStreamSplitDecode with Neon

### PR DESCRIPTION
**Pull request by guyuqi**
Improve ByteStreamSplitDecode performance for float_1k/4k/32k/64k:
1k:  21.6536G/s -> 27.7227G/s
4K:  22.1561G/s -> 28.2019G/s
32k: 21.9854G/s -> 25.8939G/s
64k: 22.0192G/s -> 26.1399G/s
No obvious performance difference for Double_1k/4k/32k/64k.

### CODE REVIEW

File: cpp/src/arrow/util/byte_stream_split.h [hunk 1]
```diff
@@ -25,23 +25,44 @@
...
 
-#if defined(ARROW_HAVE_SSE4_2)
+#if defined(ARROW_HAVE_NEON)
+#define DATA_TYPE_128BIT uint8x16_t
+#define LDR_SIMD_128BIT vld1q_u8
+#define STR_SIMD_128BIT vst1q_u8
+#define REINT_CAST_TYPE uint8_t
+#define SHUFFLE_LO_128BIT vzip1q_u8
+#define SHUFFLE_HI_128BIT vzip2q_u8
+#elif defined(ARROW_HAVE_SSE4_2)
```
Comment (cyb70289) [inline cpp/src/arrow/util/byte_stream_split.h:47]: An SIMD wrapper lib looks better than doing this ourselves. And I think there are other SSE4 codes in Arrow have similar NEON implementation.
@pitrou @jianxind

Comment (pitrou) [reply cpp/src/arrow/util/byte_stream_split.h:47]: I agree that we should discuss adopting an SIMD library. There are several of them.
https://github.com/agenium-scale/nsimd
https://github.com/aff3ct/MIPP
https://xsimd.readthedocs.io/en/latest/
https://github.com/p12tic/libsimdpp

Do you want to start the discussion on the mailing-list @cyb70289 ?

Comment (cyb70289) [reply cpp/src/arrow/util/byte_stream_split.h:47]: I've sent a mail to dev mailing-list for discussion.

**Comment by guyuqi** [comment]
From discussion on dev mailing-list,  the soft consensus among us is that adopting [xsimd](https://github.com/xtensor-stack/xsimd) as the SIMD wrapper library for Arrow. 

I'm going to contribute the missing functions to xsimd first.
So convert to draft.

**Comment by pitrou** [comment]
Thank you @guyuqi !

**Comment by pitrou** [comment]
I opened https://issues.apache.org/jira/browse/ARROW-11744 to track xsimd vendoring.

**Comment by guyuqi** [comment]
Refactor Decoding-Bytestream implementation with XSIMD
Improve ByteStreamSplitDecode performance by leveraging simd library:xsimd.

The tests of Byte-streaming-decode(parquet-internals-test)are passed on x86/Arm64:
```
[----------] 4 tests from TestByteStreamSplitEncoding/0, where TypeParam = parquet::PhysicalType<(parquet::Type::type)4>
[ RUN      ] TestByteStreamSplitEncoding/0.BasicRoundTrip
[       OK ] TestByteStreamSplitEncoding/0.BasicRoundTrip (0 ms)
[ RUN      ] TestByteStreamSplitEncoding/0.RoundTripSingleElement
[       OK ] TestByteStreamSplitEncoding/0.RoundTripSingleElement (0 ms)
[ RUN      ] TestByteStreamSplitEncoding/0.CheckOnlyDecode
[       OK ] TestByteStreamSplitEncoding/0.CheckOnlyDecode (0 ms)
[ RUN      ] TestByteStreamSplitEncoding/0.CheckOnlyEncode
[       OK ] TestByteStreamSplitEncoding/0.CheckOnlyEncode (0 ms)
[----------] 4 tests from TestByteStreamSplitEncoding/0 (0 ms total)

[----------] 4 tests from TestByteStreamSplitEncoding/1, where TypeParam = parquet::PhysicalType<(parquet::Type::type)5>
[ RUN      ] TestByteStreamSplitEncoding/1.BasicRoundTrip
[       OK ] TestByteStreamSplitEncoding/1.BasicRoundTrip (0 ms)
[ RUN      ] TestByteStreamSplitEncoding/1.RoundTripSingleElement
[       OK ] TestByteStreamSplitEncoding/1.RoundTripSingleElement (0 ms)
[ RUN      ] TestByteStreamSplitEncoding/1.CheckOnlyDecode
[       OK ] TestByteStreamSplitEncoding/1.CheckOnlyDecode (0 ms)
[ RUN      ] TestByteStreamSplitEncoding/1.CheckOnlyEncode
[       OK ] TestByteStreamSplitEncoding/1.CheckOnlyEncode (0 ms)
[----------] 4 tests from TestByteStreamSplitEncoding/1 (0 ms total)

[----------] 1 test from ByteStreamSplitEncodeDecode
[ RUN      ] ByteStreamSplitEncodeDecode.InvalidDataTypes
[       OK ] ByteStreamSplitEncodeDecode.InvalidDataTypes (0 ms)
[----------] 1 test from ByteStreamSplitEncodeDecode (0 ms total)
```

File: cpp/src/arrow/util/byte_stream_split.h [hunk 2]
```diff
@@ -23,25 +23,33 @@
 #include <stdint.h>
 #include <algorithm>
 
+#if defined(ARROW_HAVE_NEON) || defined(ARROW_HAVE_SSE4_2)
+#include <xsimd/xsimd.hpp>
+// Enable the SIMD for ByteStreamSplit Decoder
+#define ARROW_HAVE_SIMD_DECODE_SPLIT
+#endif
+
 #ifdef ARROW_HAVE_SSE4_2
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:32]: Why isn't NEON enabled here?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 3]
```diff
@@ -23,25 +23,33 @@
...
 namespace arrow {
 namespace util {
 namespace internal {
 
-#if defined(ARROW_HAVE_SSE4_2)
+#ifdef ARROW_HAVE_SIMD_DECODE_SPLIT
 template <typename T>
-void ByteStreamSplitDecodeSse2(const uint8_t* data, int64_t num_values, int64_t stride,
-                               T* out) {
+void ByteStreamSplitDecode128bit(const uint8_t* data, int64_t num_values, int64_t stride,
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:43]: Nit: call this `ByteStreamSplitDecodeSIMD128`?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 4]
```diff
@@ -63,30 +71,34 @@ void ByteStreamSplitDecodeSse2(const uint8_t* data, int64_t num_values, int64_t
...
+              &output_data[(i * kNumStreams + j) * dTypeSize128bits]),
+          stage[kNumStreamsLog2][j], xsimd::unaligned_mode());
     }
   }
 }
+#endif  // ARROW_HAVE_SIMD_DECODE_SPLIT
 
+#if defined(ARROW_HAVE_SSE4_2)
 template <typename T>
 void ByteStreamSplitEncodeSse2(const uint8_t* raw_values, const size_t num_values,
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:103]: Can you please convert this to use xsimd as well?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 5]
```diff
@@ -23,25 +23,33 @@
...
 
-#if defined(ARROW_HAVE_SSE4_2)
+#ifdef ARROW_HAVE_SIMD_DECODE_SPLIT
 template <typename T>
-void ByteStreamSplitDecodeSse2(const uint8_t* data, int64_t num_values, int64_t stride,
-                               T* out) {
+void ByteStreamSplitDecode128bit(const uint8_t* data, int64_t num_values, int64_t stride,
+                                 T* out) {
+  using simd_batch = xsimd::batch<int8_t, 16>;
+  constexpr size_t dTypeSize128bits = sizeof(int8_t) * 16;
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:46]: Can you call this `kBatchSize`?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 6]
```diff
@@ -63,30 +71,34 @@ void ByteStreamSplitDecodeSse2(const uint8_t* data, int64_t num_values, int64_t
...
-  __m128i stage[kNumStreamsLog2 + 1U][kNumStreams];
+  simd_batch stage[kNumStreamsLog2 + 1U][kNumStreams];
   constexpr size_t kNumStreamsHalf = kNumStreams / 2U;
 
   for (int64_t i = 0; i < num_blocks; ++i) {
     for (size_t j = 0; j < kNumStreams; ++j) {
-      stage[0][j] = _mm_loadu_si128(
-          reinterpret_cast<const __m128i*>(&data[i * sizeof(__m128i) + j * stride]));
+      stage[0][j] = simd_batch(
+          reinterpret_cast<const int8_t*>(&data[i * dTypeSize128bits + j * stride]),
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:80]: Is the `reinterpret_cast` still needed here?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 7]
```diff
@@ -63,30 +71,34 @@ void ByteStreamSplitDecodeSse2(const uint8_t* data, int64_t num_values, int64_t
...
-            _mm_unpackhi_epi8(stage[step][j], stage[step][kNumStreamsHalf + j]);
+            xsimd::zip_hi(stage[step][j], stage[step][kNumStreamsHalf + j]);
       }
     }
     for (size_t j = 0; j < kNumStreams; ++j) {
-      _mm_storeu_si128(reinterpret_cast<__m128i*>(
-                           &output_data[(i * kNumStreams + j) * sizeof(__m128i)]),
-                       stage[kNumStreamsLog2][j]);
+      xsimd::store_simd<int8_t, int8_t>(
+          reinterpret_cast<int8_t*>(
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:93]: Is the `reinterpret_cast` still needed here?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 8]
```diff
@@ -63,30 +71,34 @@ void ByteStreamSplitDecodeSse2(const uint8_t* data, int64_t num_values, int64_t
...
         stage[step + 1U][j * 2 + 1U] =
-            _mm_unpackhi_epi8(stage[step][j], stage[step][kNumStreamsHalf + j]);
+            xsimd::zip_hi(stage[step][j], stage[step][kNumStreamsHalf + j]);
       }
     }
     for (size_t j = 0; j < kNumStreams; ++j) {
-      _mm_storeu_si128(reinterpret_cast<__m128i*>(
-                           &output_data[(i * kNumStreams + j) * sizeof(__m128i)]),
-                       stage[kNumStreamsLog2][j]);
+      xsimd::store_simd<int8_t, int8_t>(
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:92]: I don't think you need to pass the template parameters explicitly?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 9]
```diff
@@ -63,30 +71,34 @@ void ByteStreamSplitDecodeSse2(const uint8_t* data, int64_t num_values, int64_t
...
+          reinterpret_cast<int8_t*>(
+              &output_data[(i * kNumStreams + j) * dTypeSize128bits]),
+          stage[kNumStreamsLog2][j], xsimd::unaligned_mode());
     }
   }
 }
+#endif  // ARROW_HAVE_SIMD_DECODE_SPLIT
 
+#if defined(ARROW_HAVE_SSE4_2)
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:101]: Why not `ARROW_HAVE_SIMD_ENCODE_SPLIT`?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 10]
```diff
@@ -545,21 +557,23 @@ void ByteStreamSplitEncodeAvx512(const uint8_t* raw_values, const size_t num_val
...
 template <typename T>
 void inline ByteStreamSplitDecodeSimd(const uint8_t* data, int64_t num_values,
                                       int64_t stride, T* out) {
 #if defined(ARROW_HAVE_AVX512)
   return ByteStreamSplitDecodeAvx512(data, num_values, stride, out);
 #elif defined(ARROW_HAVE_AVX2)
   return ByteStreamSplitDecodeAvx2(data, num_values, stride, out);
-#elif defined(ARROW_HAVE_SSE4_2)
-  return ByteStreamSplitDecodeSse2(data, num_values, stride, out);
+#elif defined(ARROW_HAVE_SSE4_2) || defined(ARROW_HAVE_NEON)
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:568]: Why not `ARROW_HAVE_SIMD_ENCODE_SPLIT`?

File: cpp/src/arrow/util/byte_stream_split.h [hunk 11]
```diff
@@ -545,21 +557,23 @@ void ByteStreamSplitEncodeAvx512(const uint8_t* raw_values, const size_t num_val
...
 
-#if defined(ARROW_HAVE_SIMD_SPLIT)
+#if defined(ARROW_HAVE_SIMD_DECODE_SPLIT)
 template <typename T>
 void inline ByteStreamSplitDecodeSimd(const uint8_t* data, int64_t num_values,
                                       int64_t stride, T* out) {
 #if defined(ARROW_HAVE_AVX512)
   return ByteStreamSplitDecodeAvx512(data, num_values, stride, out);
 #elif defined(ARROW_HAVE_AVX2)
   return ByteStreamSplitDecodeAvx2(data, num_values, stride, out);
```
Comment (pitrou) [inline cpp/src/arrow/util/byte_stream_split.h:567]: Can the AVX versions also be converted to use xsimd?

File: cpp/src/parquet/encoding_benchmark.cc [hunk 12]
```diff
@@ -371,17 +371,22 @@ BENCHMARK(BM_ByteStreamSplitDecode_Double_Scalar)->Range(MIN_RANGE, MAX_RANGE);
 BENCHMARK(BM_ByteStreamSplitEncode_Float_Scalar)->Range(MIN_RANGE, MAX_RANGE);
 BENCHMARK(BM_ByteStreamSplitEncode_Double_Scalar)->Range(MIN_RANGE, MAX_RANGE);
 
-#if defined(ARROW_HAVE_SSE4_2)
-static void BM_ByteStreamSplitDecode_Float_Sse2(benchmark::State& state) {
+#if defined(ARROW_HAVE_SSE4_2) || defined(ARROW_HAVE_NEON)
+static void BM_ByteStreamSplitDecode_Float_128bit(benchmark::State& state) {
```
Comment (pitrou) [inline cpp/src/parquet/encoding_benchmark.cc:375]: Perhaps "SIMD128" rather than "128bit".

**Review by pitrou — CHANGES_REQUESTED** [review]
Thanks for doing this. It would be nice if we could entirely convert this to xsimd.

File: cpp/src/arrow/util/byte_stream_split.h (hunk 2 above)
Comment (guyuqi) [reply cpp/src/arrow/util/byte_stream_split.h:32]: Hi pitrou, thanks for your comments.

I splited `ARROW_HAVE_SIMD_SPLIT` into `ARROW_HAVE_SIMD_DECODE_SPLIT` and `ARROW_HAVE_SIMD_ENCODE_SPLIT` which are used for `ByteStreamSplitDecode` and `ByteStreamSplitEncode` respectively.
This PR is for ByteStreamSplitDecode, so not enabling `Encode ` here for Neon.

How about to enable `Encode` and convert `ByteStreamSplitEncode` to xsimd in another PR and Jira? 
Or, do you prefer putting everything in this single PR? 
Thanks. :)

File: cpp/src/arrow/util/byte_stream_split.h (hunk 3 above)
Comment (guyuqi) [reply cpp/src/arrow/util/byte_stream_split.h]: Agreed, updated

File: cpp/src/arrow/util/byte_stream_split.h (hunk 5 above)
Comment (guyuqi) [reply cpp/src/arrow/util/byte_stream_split.h]: Updated, thanks.

File: cpp/src/arrow/util/byte_stream_split.h (hunk 7 above)
Comment (guyuqi) [reply cpp/src/arrow/util/byte_stream_split.h]: Yes

File: cpp/src/arrow/util/byte_stream_split.h (hunk 8 above)
Comment (guyuqi) [reply cpp/src/arrow/util/byte_stream_split.h]: It makes sense, thanks.

File: cpp/src/arrow/util/byte_stream_split.h (hunk 9 above)
Comment (guyuqi) [reply cpp/src/arrow/util/byte_stream_split.h]: Updated.

File: cpp/src/arrow/util/byte_stream_split.h (hunk 10 above)
Comment (guyuqi) [reply cpp/src/arrow/util/byte_stream_split.h:566]: This branch is included in `#if defined(ARROW_HAVE_SIMD_DECODE_SPLIT)`.
IMO, it seems we shoud separated them by ARCH Macro.

File: cpp/src/arrow/util/byte_stream_split.h (hunk 2 above)
Comment (pitrou) [reply cpp/src/arrow/util/byte_stream_split.h:32]: Sorry for the delay. It would be more logical to put everything in this single PR, IMHO.

**Comment by kszucs** [comment]
cc @pitrou

**Pull request closed without merging**